### PR TITLE
post-receive hook

### PR DESCRIPTION
**Pull request by battlmonstr**
Similar to https://github.com/dulwich/dulwich/pull/214 , but for running post-receive scripts.
What needs to be done here?

### CODE REVIEW

File: dulwich/server.py [hunk 1]
```diff
@@ -1018,6 +1018,14 @@ def handle(self):
         # backend can now deal with this refs and read a pack using self.read
         status = self._apply_pack(client_refs)
 
+        hook = self.repo.hooks.get('post-receive', None)
+        if hook:
+            hook.execute(
+                stdin='\n'.join([' '.join(i) for i in client_refs])
```
Comment (jelmer) [inline dulwich/server.py]: Please just pass in the refs; the hook can deal with formatting them for the shell script.

This makes it easier for non-shell hooks to be implemented.

File: dulwich/hooks.py [hunk 2]
```diff
@@ -160,3 +160,28 @@ def clean_msg(success, *args):
 
         ShellHook.__init__(self, 'commit-msg', filepath, 1,
                            prepare_msg, clean_msg, controldir)
+
+class PostReceiveShellHook(ShellHook):
```
Comment (jelmer) [inline dulwich/hooks.py]: You'll need an extra empty line above the class - this is probably why the tests are currently failing.

**Comment by battlmonstr** [comment]
Fixed your comments and made small improvements to the code.

**Comment by battlmonstr** [comment]
> ./dulwich/hooks.py:196:80: E501 line too long (80 > 79 characters)
Nooooo!! 😱

**Comment by battlmonstr** [comment]
@jelmer I have looked at your comments in #214 and made further improvements, please take a look.

**Comment by battlmonstr** [comment]
Woohoo, thanks @jelmer ! 🎉

**Pull request merged**